Add a common way to get NativeImage from a VideoFrame (VideoFrame::copyNativeImage())

kkinnunen-apple opened the pull request:
#### ac871ed0a668bd6fdd43e8590039b83fb3fd5d3e
<pre>
Add a common way to get NativeImage from a VideoFrame (VideoFrame::copyNativeImage())
<a href="https://bugs.webkit.org/show_bug.cg?id=302527">https://bugs.webkit.org/show_bug.cg?id=302527</a>
<a href="https://rdar.apple.com/164713414">rdar://164713414</a>

Reviewed by Youenn Fablet.

In order to use NativeImage and VideoFrame as the basic building blocks
of WebCore code, add a polymorphic VideoFrame::copyNativeImage() for
converting VideoFrame to a NativeImage.

Previously the conversion was done in ad hoc manner in cases where
it could be done, e.g. where the code had some sort of unrelated context
object such as RemoteMediaPlayerManager, AudioVideoRendererRemote,
WebGL context or WebGPU queue. Those object do not have much to do
with a particular VideoFrame, rather they&apos;re the target sinks of
the operations.

Local VideoFrame types, VideoFrameCV, VideoFrameGStreamer,
VideoFrameLibWebRTC have the data and are able to construct the
NativeImage directly. Use the same code as before. Later on, the
construction could be optimized for speed and memory use.

Remote VideoFrame type, RemoteVideoFrameProxy, holds the correct context
object for the source data, the RemoteVideoFrameObjectHeapProxy that
connects the WebContent process side reference to GPU process side
actual object. Use this to construct the NativeImage, same code as
before but invoke it directly.

Fix RemoteVideoFrameObjectHeapProxy instances to preserve across
GPUP restarts, because the RemoteVideoFrameProxy objects themselves
are persistent, held by WebCore object graphs as VideoFrame.

Later on, the holding functionality is will be moved to
RemoteSharedResourceCacheProxy. This will allow reference-only
conversion, e.g. getting RemoteNativeImageProxy out of
RemoteVideoFrameProxy.

This is work towards being able to use NativeImage as the basic
abstraction of image draw source, while not needing to transfer
the data between WCP and GPUP.

* Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
(WebCore::getImageBytesFromVideoFrame):
(WebCore::imageBytesForSource):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUQueueImpl.cpp:
(WebCore::WebGPU::QueueImpl::getNativeImage): Deleted.
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUQueueImpl.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUQueue.h:
* Source/WebCore/Modules/mediastream/ImageCapture.cpp:
(WebCore::createImageBitmap):
(WebCore::createImageBitmapViaDrawing): Deleted.
(WebCore::createImageBitmapFromNativeImage): Deleted.
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::texImageSource):
* Source/WebCore/platform/MediaStrategy.h:
(WebCore::MediaStrategy::nativeImageFromVideoFrame): Deleted.
* Source/WebCore/platform/VideoFrame.h:
* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.cpp:
(WebCore::BifurcatedGraphicsContext::drawVideoFrame):
* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.h:
* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::drawVideoFrame):
* Source/WebCore/platform/graphics/GraphicsContext.h:
* Source/WebCore/platform/graphics/GraphicsContextGL.cpp:
(WebCore::GraphicsContextGL::videoFrameToImage): Deleted.
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/NullGraphicsContext.h:
* Source/WebCore/platform/graphics/cairo/CairoOperationRecorder.cpp:
(WebCore::Cairo::OperationRecorder::drawVideoFrame):
* Source/WebCore/platform/graphics/cairo/CairoOperationRecorder.h:
* Source/WebCore/platform/graphics/cv/VideoFrameCV.h:
* Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
(WebCore::VideoFrameCV::copyNativeImage const):
(WebCore::VideoFrame::draw): Deleted.
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::drawVideoFrame):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp:
(WebCore::VideoFrame::copyNativeImage const):
(WebCore::VideoFrame::draw): Deleted.
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.h:
* Source/WebCore/platform/mediastream/libwebrtc/VideoFrameLibWebRTC.cpp:
(WebCore::VideoFrameLibWebRTC::copyNativeImage const):
* Source/WebCore/platform/mediastream/libwebrtc/VideoFrameLibWebRTC.h:
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::sharedResourceCache):
(WebKit::GPUProcessConnection::videoFrameObjectHeapProxy):
(WebKit::GPUProcessConnection::protectedVideoFrameObjectHeapProxy):
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.h:
* Source/WebKit/WebProcess/GPU/RemoteSharedResourceCacheProxy.cpp:
(WebKit::RemoteSharedResourceCacheProxy::gpuProcessConnectionDidBecomeAvailable):
(WebKit::RemoteSharedResourceCacheProxy::videoFrameObjectHeapProxy):
* Source/WebKit/WebProcess/GPU/RemoteSharedResourceCacheProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::videoFrameToImage): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextProxy.cpp:
(WebKit::RemoteGraphicsContextProxy::drawVideoFrame):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteQueueProxy.cpp:
(WebKit::WebGPU::RemoteQueueProxy::RemoteQueueProxy):
(WebKit::WebGPU::RemoteQueueProxy::getNativeImage): Deleted.
(WebKit::WebGPU::RemoteQueueProxy::protectedVideoFrameObjectHeapProxy const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteQueueProxy.h:
* Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.cpp:
(WebKit::AudioVideoRendererRemote::currentNativeImage const):
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerManager.cpp:
(WebKit::RemoteMediaPlayerManager::setUseGPUProcess):
* Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.cpp:
(WebKit::RemoteVideoFrameProxy::RemoteVideoFrameProxy):
(WebKit::m_videoFrameObjectHeapProxy):
(WebKit::RemoteVideoFrameProxy::copyNativeImage const):
(WebKit::RemoteVideoFrameProxy::pixelBuffer const):
(WebKit::m_pixelFormat): Deleted.
* Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.h:
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.cpp:
(WebKit::WebMediaStrategy::nativeImageFromVideoFrame): Deleted.
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.h:
* Source/WebKit/WebProcess/GPU/media/cocoa/MediaPlayerPrivateRemoteCocoa.mm:
(WebKit::MediaPlayerPrivateRemote::nativeImageForCurrentTime):
* Source/WebKit/WebProcess/GPU/webrtc/RemoteVideoFrameObjectHeapProxy.h:
(WebKit::RemoteVideoFrameObjectHeapProxy::create):
(WebKit::RemoteVideoFrameObjectHeapProxy::RemoteVideoFrameObjectHeapProxy):
(WebKit::RemoteVideoFrameObjectHeapProxy::gpuProcessConnectionDidBecomeAvailable):
* Source/WebKit/WebProcess/GPU/webrtc/RemoteVideoFrameObjectHeapProxyProcessor.cpp:
(WebKit::RemoteVideoFrameObjectHeapProxyProcessor::create):
(WebKit::RemoteVideoFrameObjectHeapProxyProcessor::RemoteVideoFrameObjectHeapProxyProcessor):
(WebKit::RemoteVideoFrameObjectHeapProxyProcessor::gpuProcessConnectionDidBecomeAvailable):
(WebKit::RemoteVideoFrameObjectHeapProxyProcessor::gpuProcessConnectionDidClose):
(WebKit::RemoteVideoFrameObjectHeapProxyProcessor::connection):
(WebKit::RemoteVideoFrameObjectHeapProxyProcessor::getNativeImage):
(WebKit::RemoteVideoFrameObjectHeapProxyProcessor::initialize): Deleted.
* Source/WebKit/WebProcess/GPU/webrtc/RemoteVideoFrameObjectHeapProxyProcessor.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::ensureGPUProcessConnection):
(WebKit::WebProcess::gpuProcessSharedResourceCache):
* Source/WebKit/WebProcess/WebProcess.h:

Canonical link: <a href="https://commits.webkit.org/303533@main">https://commits.webkit.org/303533@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c108e5ced4c96edddd0bf61f0558ea17d9f939dc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132735 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5230 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43810 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140264 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84762 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ada9b8cd-df78-4e83-9ccd-6c10d0429b35) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134605 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/5499 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4989 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/101483 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/68776 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d9e5275b-87e7-4e50-99cf-5cc2f3cc4905) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135681 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3872 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118910 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82275 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d0b655af-fb76-48df-a578-719d402a8318) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3758 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/1484 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83500 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/112723 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/37026 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142920 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4900 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/37613 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/109859 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4982 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/4237 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110036 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27891 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3743 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115181 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58385 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4954 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33529 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4792 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68405 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5045 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4911 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski reviewed (commented):
r+ from WebGPU side - will let someone on @WebKit/media-reviewers approve

youennf commented on Source/WebCore/platform/mediastream/libwebrtc/VideoFrameLibWebRTC.cpp:
```diff
@@ -82,6 +85,13 @@ CVPixelBufferRef VideoFrameLibWebRTC::pixelBuffer() const
+{
+    // FIXME: It is not efficient to create a conformer everytime. We might want to make it more efficient, for instance by storing it in GraphicsContext.
+    auto conformer = makeUnique<PixelBufferConformerCV>(kCVPixelFormatType_32BGRA);
+    return NativeImage::create(conformer->createImageFromPixelBuffer(protectedPixelBuffer().get()));
```
This is a bit inefficient as we create a CVPixelBuffer from a webrtc::VideoFrameBuffer and then convert it to BGRA.
We could probably directly do the conversion from webrtc::VideoFrameBuffer to BGRA.
Could you add a FIXME and file a bug?

kkinnunen-apple replied:
Kind of. Added instead:  https://bugs.webkit.org/show_bug.cgi?id=302866 VideoFrameLibWebRTC should construct its CVPixelBuffer from the LibWebRTC buffer

youennf commented on Source/WebKit/WebProcess/GPU/RemoteSharedResourceCacheProxy.h:
```diff
@@ -41,8 +42,17 @@ class RemoteSharedResourceCacheProxy : public ThreadSafeRefCounted<RemoteSharedR
 private:
     RemoteSharedResourceCacheProxy();
+#if ENABLE(VIDEO)
+    RefPtr<RemoteVideoFrameObjectHeapProxy> m_videoFrameObjectHeapProxy;
```
Could we make it a const RefPtr and use lazyInitialize?

kkinnunen-apple replied:
No, I'd not want to go there. lazyInitialize is undefined behaviour. I'll confirm in slack but I think the whole function should just be removed.

youennf commented on Source/WebKit/WebProcess/GPU/RemoteSharedResourceCacheProxy.cpp:
```diff
@@ -42,6 +46,27 @@ RemoteSharedResourceCacheProxy::RemoteSharedResourceCacheProxy() = default;
+{
+    UNUSED_PARAM(gpuProcessConnection);
+#if ENABLE(VIDEO)
+    if (m_videoFrameObjectHeapProxy)
```
We tend to write things like `if (RefPtr proxy = m_....)`.
Though we could make m_videoFrameObjectHeapProxy const to remove the refing.

youennf commented on Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.cpp:
```diff
@@ -112,22 +115,31 @@ uint32_t RemoteVideoFrameProxy::pixelFormat() const
+    // FIXME: This will change in the future to create GPUP side reference.
+#if PLATFORM(COCOA)
+    Ref frame = m_baseVideoFrame ? *m_baseVideoFrame : const_cast<RemoteVideoFrameProxy&>(*this);
+    return m_videoFrameObjectHeapProxy->getNativeImage(frame.get());
```
Is there a chance we call copyNativeImage() several times?
Should we keep a reference to the native image like we do for pixelBuffer?

kkinnunen-apple replied:
Yes, kind of. But not for this patch.
The force is that most frames end up duplicating the frame data, and we need to have a mitigation for this. It will come later.

youennf commented on Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.h:
```diff
@@ -90,8 +91,7 @@ class RemoteVideoFrameProxy final : public WebCore::VideoFrame {
     uint32_t m_pixelFormat { 0 };
-    // FIXME: Remove this.
-    mutable RefPtr<RemoteVideoFrameObjectHeapProxy> m_videoFrameObjectHeapProxy;
+    Ref<RemoteVideoFrameObjectHeapProxy> m_videoFrameObjectHeapProxy;
```
```suggestion
    const Ref<RemoteVideoFrameObjectHeapProxy> m_videoFrameObjectHeapProxy;
```

youennf commented on Source/WebKit/WebProcess/GPU/webrtc/RemoteVideoFrameObjectHeapProxyProcessor.cpp:
```diff
@@ -60,28 +56,43 @@ RemoteVideoFrameObjectHeapProxyProcessor::~RemoteVideoFrameObjectHeapProxyProces
 {
-    RefPtr<IPC::Connection> connection;
+    gpuProcessConnection.addClient(*this);
+    RefPtr<IPC::Connection> connection = gpuProcessConnection.connection();
```
```suggestion
    Ref connection = gpuProcessConnection.connection();
```

youennf commented on Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
```diff
@@ -211,14 +209,12 @@ void GPUProcessConnection::resetAudioMediaStreamTrackRendererInternalUnit(AudioM
-    if (!m_videoFrameObjectHeapProxy)
-        m_videoFrameObjectHeapProxy = RemoteVideoFrameObjectHeapProxy::create(*this);
-    return *m_videoFrameObjectHeapProxy;
+    return sharedResourceCache().videoFrameObjectHeapProxy();
```
It seems that the `videoFrameObjectHeapProxy` might span multiple GPU processes.
What happens in that case:
- A VideoFrame V1 is created in GPUProcess G1 and we have its remote in WebProcess.
- G1 dies and G2 is created.
- We try to get V1 data from G2.

Previously, this would be guaranteed to fail, as G1 is gone, so is V1 data.
With this patch, is there a possibility that G2 confuses V1 with another video frame it created and pass wrong data?

kkinnunen-apple replied:
It's going to fail as the new connection representation in GPUP side doesn't have the objects by the identifiers.

Note: this is existing behavior for the most part.
E.g. if you look at the previous Processor::getNativeImage, it re-establishes the connection after crash.

Later on this will be made more rigorous and conforming to the other GPUP behavior by making WCP select the identifiers and the shared cache keeping a bit better track of what's being sent to GPUP.

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/webrtc/RemoteVideoFrameObjectHeapProxyProcessor.cpp:
```diff
@@ -150,11 +161,12 @@ void RemoteVideoFrameObjectHeapProxyProcessor::newConvertedVideoFrameBuffer(std:
 
 RefPtr<NativeImage> RemoteVideoFrameObjectHeapProxyProcessor::getNativeImage(const WebCore::VideoFrame& videoFrame)
 {
-    Ref connection = WebProcess::singleton().ensureGPUProcessConnection().connection();
```
Previous similar behavior here: even though the GPUP crashed and new objects get reference to new processor, the old instances use the old processor that will just re-establish the connection and ask for the data.

youennf commented on Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
```diff
@@ -409,12 +409,12 @@ static void clampDimension(WebGPU::Extent3D& extent3D, size_t dimension, WebGPU:
         return callback({ }, 0, 0);
 
-    RefPtr<NativeImage> nativeImage = backing.getNativeImage(*videoFrame.get());
+    RefPtr<NativeImage> nativeImage = videoFrame->copyNativeImage();
```
```suggestion
    RefPtr nativeImage = videoFrame->copyNativeImage();
```

kkinnunen-apple committed:
Add a common way to get NativeImage from a VideoFrame (VideoFrame::copyNativeImage())
https://bugs.webkit.org/show_bug.cg?id=302527
rdar://164713414

Reviewed by Youenn Fablet.

In order to use NativeImage and VideoFrame as the basic building blocks
of WebCore code, add a polymorphic VideoFrame::copyNativeImage() for
converting VideoFrame to a NativeImage.

Previously the conversion was done in ad hoc manner in cases where
it could be done, e.g. where the code had some sort of unrelated context
object such as RemoteMediaPlayerManager, AudioVideoRendererRemote,
WebGL context or WebGPU queue. Those object do not have much to do
with a particular VideoFrame, rather they're the target sinks of
the operations.

Local VideoFrame types, VideoFrameCV, VideoFrameGStreamer,
VideoFrameLibWebRTC have the data and are able to construct the
NativeImage directly. Use the same code as before. Later on, the
construction could be optimized for speed and memory use.

Remote VideoFrame type, RemoteVideoFrameProxy, holds the correct context
object for the source data, the RemoteVideoFrameObjectHeapProxy that
connects the WebContent process side reference to GPU process side
actual object. Use this to construct the NativeImage, same code as
before but invoke it directly.

Fix RemoteVideoFrameObjectHeapProxy instances to preserve across
GPUP restarts, because the RemoteVideoFrameProxy objects themselves
are persistent, held by WebCore object graphs as VideoFrame.

Later on, the holding functionality is will be moved to
RemoteSharedResourceCacheProxy. This will allow reference-only
conversion, e.g. getting RemoteNativeImageProxy out of
RemoteVideoFrameProxy.

This is work towards being able to use NativeImage as the basic
abstraction of image draw source, while not needing to transfer
the data between WCP and GPUP.

* Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
(WebCore::getImageBytesFromVideoFrame):
(WebCore::imageBytesForSource):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUQueueImpl.cpp:
(WebCore::WebGPU::QueueImpl::getNativeImage): Deleted.
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUQueueImpl.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUQueue.h:
* Source/WebCore/Modules/mediastream/ImageCapture.cpp:
(WebCore::createImageBitmap):
(WebCore::createImageBitmapViaDrawing): Deleted.
(WebCore::createImageBitmapFromNativeImage): Deleted.
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::texImageSource):
* Source/WebCore/platform/MediaStrategy.h:
(WebCore::MediaStrategy::nativeImageFromVideoFrame): Deleted.
* Source/WebCore/platform/VideoFrame.h:
* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.cpp:
(WebCore::BifurcatedGraphicsContext::drawVideoFrame):
* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.h:
* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::drawVideoFrame):
* Source/WebCore/platform/graphics/GraphicsContext.h:
* Source/WebCore/platform/graphics/GraphicsContextGL.cpp:
(WebCore::GraphicsContextGL::videoFrameToImage): Deleted.
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/NullGraphicsContext.h:
* Source/WebCore/platform/graphics/cairo/CairoOperationRecorder.cpp:
(WebCore::Cairo::OperationRecorder::drawVideoFrame):
* Source/WebCore/platform/graphics/cairo/CairoOperationRecorder.h:
* Source/WebCore/platform/graphics/cv/VideoFrameCV.h:
* Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
(WebCore::VideoFrameCV::copyNativeImage const):
(WebCore::VideoFrame::draw): Deleted.
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::drawVideoFrame):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp:
(WebCore::VideoFrame::copyNativeImage const):
(WebCore::VideoFrame::draw): Deleted.
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.h:
* Source/WebCore/platform/mediastream/libwebrtc/VideoFrameLibWebRTC.cpp:
(WebCore::VideoFrameLibWebRTC::copyNativeImage const):
* Source/WebCore/platform/mediastream/libwebrtc/VideoFrameLibWebRTC.h:
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::sharedResourceCache):
(WebKit::GPUProcessConnection::videoFrameObjectHeapProxy):
(WebKit::GPUProcessConnection::protectedVideoFrameObjectHeapProxy):
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.h:
* Source/WebKit/WebProcess/GPU/RemoteSharedResourceCacheProxy.cpp:
(WebKit::RemoteSharedResourceCacheProxy::gpuProcessConnectionDidBecomeAvailable):
(WebKit::RemoteSharedResourceCacheProxy::videoFrameObjectHeapProxy):
* Source/WebKit/WebProcess/GPU/RemoteSharedResourceCacheProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::videoFrameToImage): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextProxy.cpp:
(WebKit::RemoteGraphicsContextProxy::drawVideoFrame):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteQueueProxy.cpp:
(WebKit::WebGPU::RemoteQueueProxy::RemoteQueueProxy):
(WebKit::WebGPU::RemoteQueueProxy::getNativeImage): Deleted.
(WebKit::WebGPU::RemoteQueueProxy::protectedVideoFrameObjectHeapProxy const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteQueueProxy.h:
* Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.cpp:
(WebKit::AudioVideoRendererRemote::currentNativeImage const):
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerManager.cpp:
(WebKit::RemoteMediaPlayerManager::setUseGPUProcess):
* Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.cpp:
(WebKit::RemoteVideoFrameProxy::RemoteVideoFrameProxy):
(WebKit::m_videoFrameObjectHeapProxy):
(WebKit::RemoteVideoFrameProxy::copyNativeImage const):
(WebKit::RemoteVideoFrameProxy::pixelBuffer const):
(WebKit::m_pixelFormat): Deleted.
* Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.h:
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.cpp:
(WebKit::WebMediaStrategy::nativeImageFromVideoFrame): Deleted.
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.h:
* Source/WebKit/WebProcess/GPU/media/cocoa/MediaPlayerPrivateRemoteCocoa.mm:
(WebKit::MediaPlayerPrivateRemote::nativeImageForCurrentTime):
* Source/WebKit/WebProcess/GPU/webrtc/RemoteVideoFrameObjectHeapProxy.h:
(WebKit::RemoteVideoFrameObjectHeapProxy::create):
(WebKit::RemoteVideoFrameObjectHeapProxy::RemoteVideoFrameObjectHeapProxy):
(WebKit::RemoteVideoFrameObjectHeapProxy::gpuProcessConnectionDidBecomeAvailable):
* Source/WebKit/WebProcess/GPU/webrtc/RemoteVideoFrameObjectHeapProxyProcessor.cpp:
(WebKit::RemoteVideoFrameObjectHeapProxyProcessor::create):
(WebKit::RemoteVideoFrameObjectHeapProxyProcessor::RemoteVideoFrameObjectHeapProxyProcessor):
(WebKit::RemoteVideoFrameObjectHeapProxyProcessor::gpuProcessConnectionDidBecomeAvailable):
(WebKit::RemoteVideoFrameObjectHeapProxyProcessor::gpuProcessConnectionDidClose):
(WebKit::RemoteVideoFrameObjectHeapProxyProcessor::connection):
(WebKit::RemoteVideoFrameObjectHeapProxyProcessor::getNativeImage):
(WebKit::RemoteVideoFrameObjectHeapProxyProcessor::initialize): Deleted.
* Source/WebKit/WebProcess/GPU/webrtc/RemoteVideoFrameObjectHeapProxyProcessor.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::ensureGPUProcessConnection):
(WebKit::WebProcess::gpuProcessSharedResourceCache):
* Source/WebKit/WebProcess/WebProcess.h:

Canonical link: https://commits.webkit.org/303533@main
